### PR TITLE
Remove needless forking to GENERIC in TransportMultiSearchAction

### PR DESCRIPTION
**Pull request by original-brownbear**
Forking to `GENERIC` is unnecessary as a fix for stack-depth problems. We can instead run a different code-path, depending on whether or not a search could be dealt with right away without forking. We have seen a case where forking to generic resulted in a large number of idle `GENERIC` threads and unpredictable latencies to the point of the user noticing these latencies.
Also, the existing logic was not correct in many cases and would fork more often than necessary. Since the msearch action starts out on a transport thread, we might very well receive the response to an individual search on that same thread, but not within the same call-stack, simply as a result of the REST channel that triggered the search and the channel used to a data node running on the same transport thread.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 1]
```diff
@@ -134,31 +130,39 @@ void executeSearch(
...
-             * of concurrent requests. At first glance, it appears that we should never poll from the queue and not obtain a request given
-             * that we only poll here no more times than the number of requests. However, this is not the only consumer of this queue as
-             * earlier requests that have already completed will poll from the queue too and they could complete before later polls are
-             * invoked here. Thus, it can be the case that we poll here and the queue was empty.
-             */
-            return;
+        // If we have another request to execute, we execute it. If the execution forked #doExecuteSearch will return false and will
+        // recursively call this method again eventually. If it did not fork and was able to execute the search right away #doExecuteSearch
+        // will return true, in which case we continue and run the next search request here.
+        while (request != null && doExecuteSearch(requests, responses, responseCounter, relativeStartTime, request, listener)) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:144]: There's a mild change in behaviour here. We will now potentially run slightly longer tasks on the transport threads if we have a bunch of search requests that get a response right away back to back. It seems to me this is a rare scenario and if encountered, then it also is quite cheap on a per-request basis.
=> It seemed fine to me to not worry about this.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:144]: We know for a fact that Search is sometimes blocking transport threads (especially in can_match & when there are many indices, etc.). We have other issues describing this particular kind of bug. 

Wouldn't it be prudent to fix that FIRST and then adjust this?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:144]: That is a fair point. You could argue this both ways though and go with: If there's so many non-forking searches that not forking blocks us here, then forking will result in horrible latency from essentially forking N-searches (N-concurrency-level) times relative to this solution.
Forking isn't free at all, especially to `GENERIC`, so there's always going to be a cutoff here where one or the other behaves better (the linked issue shows a O(100ms) difference here).

Lets discuss this on another channel, ping incoming there.

**Comment by benwtrent** [comment]
I do think we should backport to 8.15.x as a bugfix once we are ready for it.

File: server/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:144]: I'm starting to feel like this is good to go now that #110860 is merged. The can_match situation is a lot less relevant with that fix alone and also on second thought :D how do I even get into the situation of many indices and this thing resolving on the current thread? The indices targeted would all have to reside on the current node which is not realistic?

**Comment by original-brownbear** [comment]
> I do think we should backport to 8.15.x as a bugfix once we are ready for it.

Are you sure? There's no regression here in the classical sense, it's just there are some scenarios where we are needlessly forking too much and that only affects very low-latency use cases?

**Comment by benwtrent** [comment]
@original-brownbear 

Yes, you are correct. Since this is a longer standing bug, waiting for 8.16 is OK & it will allow us to catch other things. 

I am just trying to be more aware on what we can backport to 8.15 safely.

We can keep this just in 8.16

**Review by jdconrad — APPROVED** [review]
LGTM! Nice way to remove the generic threads.

**Review by javanna — COMMENTED** [review]
I am good with this, provided that we have adequate testing to surface potential issues with this change. Do we have coverage for the issues that the forking somehow addressed until now?

**Comment by original-brownbear** [comment]
> Do we have coverage for the issues that the forking somehow addressed until now?

@javanna jup we have `org.elasticsearch.action.search.TransportMultiSearchActionTests#testBatchExecute` that tests this. I don't think it reproduces the issue today because other improvements we made to the network layer have vastly reduced stack depth there but it definitely guards us agains stack-overflows should things change on the network layer :)

**Comment by original-brownbear** [comment]
Thanks you two!

**Pull request merged**